knowledge/knowledge_read.go

```diff
@@ -23,190 +23,184 @@ import (
-func (c *Client) ListResourcesByProperty(
+// ListIdentities is a helper function that lists all Identity nodes.
+func (c *Client) ListIdentities(
 	ctx context.Context,
-	property *knowledgepb.Property,
-	opts ...grpc.CallOption) ([]*knowledgepb.Node, error) {
-	return c.ListNodesByProperty(ctx, Resource, property, opts...)
+	returns []*knowledgepb.Return,
+	opts ...grpc.CallOption) ([]*knowledgeobjects.Node, error) {
+	return c.ListNodes(ctx, returns, true, opts...)
 }
 
-// ListDigitalTwins is a helper function that lists all DigitalTwin nodes.
-func (c *Client) ListDigitalTwins(
-	ctx context.Context,
-	opts ...grpc.CallOption) ([]*knowledgepb.Node, error) {
-	return c.ListNodes(ctx, DigitalTwin, opts...)
-}
-
-// ListResources is a helper function that lists all Resource nodes.
-// that have the specified property.
-func (c *Client) ListResources(
-	ctx context.Context,
-	opts ...grpc.CallOption) ([]*knowledgepb.Node, error) {
-	return c.ListNodes(ctx, Resource, opts...)
-}
-
-// ListNodes is a helper function that lists all nodes by type, regardless of whether they are DigitalTwins
-// or Resources. The nodeType argument should be in PascalCase.
+// ListNodes is a helper function that lists all nodes by type, regardless of whether they are Identities
+// or not.
 func (c *Client) ListNodes(
 	ctx context.Context,
-	nodeType string,
+	returns []*knowledgepb.Return,
+	isIdentity bool,
 	opts ...grpc.CallOption,
-) ([]*knowledgepb.Node, error) {
+) ([]*knowledgeobjects.Node, error) {
 	ctx = insertMetadata(ctx, c.xMetadata)
-	path := fmt.Sprintf("(:%s)", nodeType)
-	resp, err := c.client.IdentityKnowledge(ctx, &knowledgepb.IdentityKnowledgeRequest{
-		Path:      path,
-		Operation: knowledgepb.Operation_OPERATION_READ,
+	query := "MATCH (n:Resource)"
+	if isIdentity {
+		query = "MATCH (n:DigitalTwin)"
+	}
+	params := map[string]*knowledgepb.InputParam{}
+	resp, err := c.client.IdentityKnowledgeRead(ctx, &knowledgepb.IdentityKnowledgeReadRequest{
+		Query:       query,
+		InputParams: params,
+		Returns:     returns,
 	}, opts...)
 	if err != nil {
 		return nil, err
 	}
-	return parseMultipleNodesFromPaths(resp.GetPaths())
+	return ParseMultipleNodesFromNodes(resp.GetNodes())
 }
 
 // ListNodesByProperty is a helper function that lists all nodes that have the specified type and property.
 func (c *Client) ListNodesByProperty(
 	ctx context.Context,
-	nodeType string,
-	property *knowledgepb.Property,
+	property *knowledgeobjects.Property,
+	returns []*knowledgepb.Return,
+	isIdentity bool,
 	opts ...grpc.CallOption,
-) ([]*knowledgepb.Node, error) {
+) ([]*knowledgeobjects.Node, error) {
 	ctx = insertMetadata(ctx, c.xMetadata)
-	path := fmt.Sprintf("(n:%s)", nodeType)
-	conditions := fmt.Sprintf("WHERE n.%s = $%s", property.Key, property.Key)
+	query := "MATCH (n:Resource)"
+	if isIdentity {
+		query = "MATCH (n:DigitalTwin)"
+	}
+	query = fmt.Sprintf(
+		"%s WHERE n.%s=$%s",
+		query,
+		property.Type,
+		property.Type,
+	)
```

Comment:
This won't work anymore (it does right now, but only because we are temporarily backwards compatible with the old property model). You need to look for the property node
```
MATCH (n:Resource)-[:HAS]->(p:Property)
WHERE p.type = $type AND p.value = $value
```
and then the appropriate input params

Reply:
updated